bindings/rust/standard/integration/src/mtls.rs

```diff
@@ -0,0 +1,522 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+// SPDX-License-Identifier: Apache-2.0
+
+// This test suite exercises mTLS interoperability between s2n-tls and rustls,
+// including:
+//   - basic mTLS handshakes (TLS 1.2 and 1.3)
+//   - sync certificate validation callbacks
+//   - async certificate validation callbacks wired directly via the C FFI
+//
+// Async callbacks are registered via s2n_tls_sys instead of the Rust bindings
+// to avoid exposing an unstable async callback API in the public Rust surface.
+
+use std::{
+    mem,
+    os::raw::c_void,
+    ptr::NonNull,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        mpsc::{Receiver, Sender},
+        Arc,
+    },
+};
+
+use rustls::ClientConfig;
+
+use s2n_tls::{
+    callbacks::{CertValidationCallbackSync, CertValidationInfo, VerifyHostNameCallback},
+    config::{Builder, Config},
+    connection::Connection,
+    enums::ClientAuthType,
+    error::Error as S2NError,
+};
+
+use s2n_tls_sys::{
+    s2n_cert_validation_accept, s2n_cert_validation_info, s2n_config,
+    s2n_config_set_cert_validation_cb, s2n_connection, s2n_status_code,
+};
+
+use tls_harness::{
+    cohort::{RustlsConfig, RustlsConnection, S2NConfig, S2NConnection},
+    harness::{read_to_bytes, TlsConfigBuilder},
+    PemType, SigType, TlsConnPair, TlsConnection,
+};
+
+const APP_DATA_SIZE: usize = 100_000;
+
+/// A wrapper around a raw pointer to `s2n_cert_validation_info` that can be sent across threads.
+///
+/// This is used in tests to simulate async certificate validation where the validation
+/// decision is deferred and made on a different thread or after some async operation.
+struct SendableCertValidationInfo(*mut s2n_cert_validation_info);
+
+// SAFETY: The pointer is owned by s2n-tls and remains valid for the duration of the
+// pending async validation (until accept() or reject() is called, or the connection is freed).
+// The test mimics the intended usage pattern where an application hands off the pointer
+// to a worker thread that later calls accept()/reject().
+unsafe impl Send for SendableCertValidationInfo {}
+
+/// Get the raw s2n_config pointer from S2NConfig
+/// SAFETY: S2NConfig wraps Config, which is a thin NonNull<s2n_config>
+/// wrapper. This is test-only functionality relying on that internal layout.
+unsafe fn raw_config(cfg: &mut S2NConfig) -> *mut s2n_config {
+    let config: &mut Config = &mut cfg.config;
+    let nn: &mut NonNull<s2n_config> = mem::transmute(config);
+    nn.as_ptr()
+}
+
+#[derive(Debug)]
+struct TestCertValidationCallback {
+    invoked: Arc<AtomicU64>,
+    immediately_accept: bool,
+}
+
+impl TestCertValidationCallback {
+    fn new_sync() -> Self {
+        Self {
+            invoked: Arc::new(AtomicU64::new(0)),
+            immediately_accept: true,
```

Comment:
This field is always true since it's for the sync case. I think it can be removed?